### ISSUE TITLE
Setting float8 weight data type in OneTrainer increases memory usage by 50%

### ISSUE THREAD

**Issue by mcmanus82** [LoganBooker/prodigy-plus-schedule-free#20]
Hi
I did a non public implementation for you optimizer for OneTrainer because i really liked it in Kohya_SS(GUI). I did not submit it yet because i wanted to see if i can make it work.

Everything seems to work alright, but when i change the training type from fp16 to 8 the memory usage increases by 50%, instead of going down.(8Gb->12Gb)

I tried disabling Factored_32 but it made no difference.

I have attached the implementation to OneTrainer as a zip(only the files i modified) it is for the latest OneTrainer and 1.9.0 Prodigy+

[OT-ProdyPlus.zip](https://github.com/user-attachments/files/19621965/OT-ProdyPlus.zip)

**Comment by LoganBooker** [comment]
Hi mcmanus82, thanks for your interest. The optimiser is mainly aimed at bf16/fp32 training (and mixes thereof), so I can't make any assurances about its stability or suitability with other precisions.

That said, my inclination in regards to your problem is likely this line, which forces the dtype of some of the state variables to be bf16 or fp32: https://github.com/LoganBooker/prodigy-plus-schedule-free/blob/main/prodigyplus/core_optimiser.py#L199

You can try changing this to just use the dtype of the gradient, but again, I'm not sure how this will affect the performance of the optimiser.

**Comment by LoganBooker** [comment]
Hi mcmanus82, thanks for your interest. The optimiser is aimed at bf16 and/or fp32 training, so I can't provide any assurances about its suitability with other precisions (or whether it's compatible at all).

As for your specific issue, I can't say exactly why memory increases at fp8. It's possible there's some overhead happening somewhere outside of the optimiser, perhaps related to autocast, but I don't have enough knowledge of OneTrainer or PyTorch to offer help.

If you're using mixed precision, it might be worth checking that the state initialisation is using the dtype you want:
https://github.com/LoganBooker/prodigy-plus-schedule-free/blob/d6781212761d57ea2ef2cf841dbc3f63003f27c1/prodigyplus/core_optimiser.py#L193-L237

The optimiser uses the gradient's dtype, but if that somehow differs from the weight dtype, it's possible the size inconsistency is happening there.

**Comment by mcmanus82** [comment]
Thanks for your reply :) i will try to figure this out as it seems to work in Kohya_SS(GUI) so it is most likely something OneTrainer does.

Another thing, after testing my implementation it turns out the optimizer only kinda works in OneTrainer. Fused back pass reduces memory and seems to function, lr is increased in a similar fashion as in Kohya, but the actual rate of learning is a lot faster. It over trains in a couple hundred steps.

I know its more of a OneTrainer question, but do you think i left something out in the implementation or that OneTrainer's workflow needs to be re-adjusted to be compatible with Prodigy+?

Thanks again.

**Comment by LoganBooker** [comment]
Hmm... the original fused backward pass implementation was actually done at the request of a OneTrainer user (https://github.com/LoganBooker/prodigy-plus-schedule-free/issues/1) where it seemed to work as intended (at least, for their hacked implementation). The only thing I can think of is to make sure the `fused_back_pass` is set to `True` in the optimiser, otherwise it's possible the optimiser will run twice -- once per gradient when the hook is called, and again for all gradients when `step` is called.

**Comment by mcmanus82** [comment]
After a day of testing i think i can describe the issue much more clearly. So basically there was a misunderstanding. In OneTrainer you can set two main "Data types". 

One is on the training tab, called "Train data type" and this is what you were referring to when you said it only supports bf16 / fp32. And it is true, if you leave this at its default value of fp16 it breaks training. I don't think anyone would mind about this requirement but a note or a warning print might be warranted.

The second data type is on the Model tab, called "Weight data type" which refers to the base model's data type to use. If you set this to float8 with bf16 for "Train data type" then it increases the VRAM usage by 50%

Contrary to my previous statement, Fused back pass does not seem to work, the only message Prodigy+ shows is:
"Optimiser contains single param_group -- 'split_groups' has been disabled."
I'm not training the text encoder(SDXL) so i presume this is normal.

Should there be a message in OneTrainer if it does Fuse back pass like in SD-Scripts? As far as i can tell it had no effect on VRAM usage, even with SDXL it should free up a couple hundred megs i guess.

**Comment by mcmanus82** [comment]
Forgot to mention this, but i'm doing a more proper implementation, as in adding Prodigy+ as a new optimizer and not replacing Prodigy.

I also added the new one to the "is_adaptive" and "supports_fused_back" groups. So it should have all the properties that the hacked previous attempt had.

**Comment by mcmanus82** [comment]
Never mind the fp8 VRAM usage, it turned out to be a OneTrainer SDXL specific bug... That only leaves the Fused back pass thing

**Comment by LoganBooker** [comment]
Glad to hear the fp8 issue was resolved -- can definitely be tricky to chase these things down when there's so many moving parts.

> "Optimiser contains single param_group -- 'split_groups' has been disabled."
> I'm not training the text encoder(SDXL) so i presume this is normal.

That's correct,

> Should there be a message in OneTrainer if it does Fuse back pass like in SD-Scripts?

No, there's no specific message for trainers other than Kohya, as the optimiser has special hooking/patching code to work correctly with it. This is actually a workaround of sorts -- Kohya should probably have more generic handling for optimisers that support FBP. OneTrainer is a little more flexible in this regard, but still requires a lot of plumbing because a lot of the logic is hardcoded (as you've noted).

I'll apply your patch with OneTrainer and see if there's anything amiss going on. But as far as I know, all that should be required is making sure `optimizer.supports_fused_back_pass()` returns true and `fused_back_pass` is set to true as part of the optimiser arguments.

**Comment by LoganBooker** [comment]
I ran a quick test with your patch, and can confirm fused backward pass works as expected. With `fused_back_pass` set to true, the optimiser skips the main `step` call, and only processes calls to `step_parameter`.

I will say that if you're training a LoRA, rather than a fine-tune, the memory savings are fairly minimal with FBP, as the gradients are significantly smaller, especially if you're training at ranks such as 8 or 16. Usually, the overhead of processing each gradient hook outweighs any memory gains.
